src/wmtk/operations/tri_mesh/VertexAttributesUpdate.cpp

```diff
@@ -0,0 +1,64 @@
+#include "VertexAttributesUpdate.hpp"
+
+#include <wmtk/SimplicialComplex.hpp>
+#include <wmtk/TriMesh.hpp>
+
+namespace wmtk::operations {
+void OperationSettings<tri_mesh::VertexAttributesUpdate>::initialize_invariants(const TriMesh& m)
+{
+    // outdated + is valid tuple
+    invariants = basic_invariant_collection(m);
+}
+
+namespace tri_mesh {
+VertexAttributesUpdate::VertexAttributesUpdate(
+    Mesh& m,
+    const Tuple& t,
+    const OperationSettings<VertexAttributesUpdate>& settings)
+    : TriMeshOperation(m)
+    , TupleOperation(settings.invariants, t)
+    , m_settings{settings}
+{}
+
+std::string VertexAttributesUpdate::name() const
+{
+    return "tri_mesh_vertex_attributes_update";
+}
+
+const Tuple& VertexAttributesUpdate::return_tuple() const
+{
+    return m_output_tuple;
+}
+
+bool VertexAttributesUpdate::before() const
+{
+    if (!mesh().is_valid_slow(input_tuple())) {
```

Comment:
This should/will be done by the invariants but there is no need to change it right now because it will be refactored anyways.